src/prefect/backend/flow_run.py

```diff
@@ -722,7 +722,7 @@ def get_all_task_runs(self) -> List["TaskRunView"]:
         task_run_data = TaskRunView._query_for_task_runs(
             where={
                 "flow_run_id": {"_eq": self.flow_run_id},
-                "task_run_id": {"_not", {"_in": list(self._cached_task_runs.keys())}},
+                "task_id": {"_nin": list(self._cached_task_runs.keys())},
```

Comment:
Thanks for catching this! I believe this should be

```suggestion
                "id": {"_nin": list(self._cached_task_runs.keys())},
```

since we're querying the task run table and excluding task run ids not task ids.

This looks right; I can confirm this works as intended locally as well since all of our tests for this mock the backend API

Reply:
Thanks Michael, a couple of things:
1) changed to `task_id` to `id`, that makes sense.
2) I'm still getting two instances for every task. eg for a 4-node flow:
```
0:TaskRunView(task_run_id='8c320aee-dc45-4345-8d7f-9473cb334317', task_id='ba55c7f3-bc7c-4a2a-911b-280a309b6567', task_slug="noop['a']-1", state=<Success: "Task run succeeded.">, result=1)
1:TaskRunView(task_run_id='de059738-656e-4dfe-8cbb-37c89ac24337', task_id='5edf0ea5-92fa-4a62-ad0f-34edde717ab1', task_slug='noop-1', state=<Success: "Task run succeeded.">, result={'a': 1})
2:TaskRunView(task_run_id='78ee8ff2-ebc5-4e7e-a4ec-5c5754a032fe', task_id='92e1bfaa-99b5-49fb-a4ba-3aa918488782', task_slug="x['a']-1", state=<Success: "Task run succeeded.">, result=1)
3:TaskRunView(task_run_id='29505fad-afb2-46d2-acca-5df47aa11ef5', task_id='593edfb8-3bbb-456c-ae05-c74503bab709', task_slug='x', state=<Success: "Task run succeeded.">, result={'a': 1})
4:TaskRunView(task_run_id='8c320aee-dc45-4345-8d7f-9473cb334317', task_id='ba55c7f3-bc7c-4a2a-911b-280a309b6567', task_slug="noop['a']-1", state=<Success: "Task run succeeded.">, result=1)
5:TaskRunView(task_run_id='de059738-656e-4dfe-8cbb-37c89ac24337', task_id='5edf0ea5-92fa-4a62-ad0f-34edde717ab1', task_slug='noop-1', state=<Success: "Task run succeeded.">, result={'a': 1})
6:TaskRunView(task_run_id='78ee8ff2-ebc5-4e7e-a4ec-5c5754a032fe', task_id='92e1bfaa-99b5-49fb-a4ba-3aa918488782', task_slug="x['a']-1", state=<Success: "Task run succeeded.">, result=1)
7:TaskRunView(task_run_id='29505fad-afb2-46d2-acca-5df47aa11ef5', task_id='593edfb8-3bbb-456c-ae05-c74503bab709', task_slug='x', state=<Success: "Task run succeeded.">, result={'a': 1})
```

3) why did the windows unit tests fail below?
4) `I can confirm this works as intended locally as well since all of our tests for this mock the backend API`: does that mean there are already tests, i.e. I don't need to add a test for this?

Reply:
2. It looks like this is because we add all the newly returned tasks to the cache then return the new tasks plus the cache. The return value should be determined before the cache is updated.
3. There's something wrong with CI. We're looking into it, you can ignore it for now.
4. The tests for this class are at `tests/backend/test_flow_run.py` but this is actually not covered by unit tests. You can look at how the other tests work there and try to write a test or I can help do so.

Reply:
`It looks like this is because we add all the newly returned tasks to the cache then return the new tasks plus the cache. The return value should be determined before the cache is updated.` 
Updated, let me know if that looks good to you. I think we should simply return the cache which at that point should be complete (with both old cached and updated tasks).